File: demo-dubbo-gradle/demo-dubbo-gradle-consumer/src/main/java/com/example/demo/dubbo/controller/UserController.java
package com.example.demo.dubbo.controller;

import com.example.demo.dubbo.IUserService;
import com.example.demo.dubbo.entity.User;
import org.apache.dubbo.config.annotation.DubboReference;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("user")
public class UserController {
    @DubboReference(protocol = "dubbo", url = "dubbo://localhost:20880")
    private IUserService userService;

    @GetMapping("{name}")
    public String get(@PathVariable String name) {
        return userService.hello(name);
    }

    @GetMapping
    public User add(User user) {
        return userService.add(user);
    }
}

File: demo-spring-boot-kotlin-gradle-protobuf/src/main/kotlin/com/example/demo/protobuf/controller/UserController.kt
package com.example.demo.protobuf.controller

import com.example.demo.protobuf.entity.User
import com.example.demo.protobuf.entity.User.AddressBook
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("user")
class UserController {
    @GetMapping(produces = ["application/x-protobuf"])
    fun get(): AddressBook {
        return AddressBook.newBuilder()
            .addPeople(User.Person.newBuilder().addPhones(User.Person.PhoneNumber.newBuilder().setNumber("1111")))
            .build()
    }
}